### ISSUE TITLE
Clarification on `Reference`

### ISSUE THREAD

**Issue by Tokazama** [JuliaSIMD/ManualMemory.jl#6]
I wanted to use `load` to take advantage of the structure loading and had a few questions on `ManualMemory.Reference`.

1. How is this different from `Base.RefValue`
2. Would it make sense to have `Reference{T}()` defined?

**Comment by chriselrod** [comment]
1. The current use case is for `StrideArraysCore.object_and_preserve`:
 https://github.com/JuliaSIMD/StrideArraysCore.jl/blob/ef25a02989f60bcfef1d2d71240654934fd13a4a/src/thread_compatible.jl#L2

```julia
obj, pres = object_and_preserve(x)
```
is used for storing data in a pointer; we store `obj` and `GC.@preserve pres begin; ...; end`
When we want to pass something by reference, `object_and_preserve` wraps it in a `ManualMemory.Reference`.

Then, the [load](https://github.com/JuliaSIMD/ManualMemory.jl/blob/1ddedc8aa4a3c71943ccd3042c767f1331a6c9d1/src/ManualMemory.jl#L54) function loads the mutable struct, and automatically dereferences it.

An example use case is that `ThreadingUtilities.jl` uses a very limited per-thread buffer for passing data.
It's not that rare for structs to be much larger than this buffer. Trixi.jl, for example, uses some very large structs.
So `object_and_preserve` can wrap these giant structs in a `Reference`.
Then when the receiver `ManualMemory.load`s the `Reference`, the `Reference` wrapper automatically gets stripped, returning the struct. Hopefully the compiler at this point is smart enough to avoid any unnecessary/extra copying.

This should also have the performance benefit of storing the large object just once (on the stack) instead of once per thread we're passing it to (in their respective buffers).

So why use `ManualMemory.Reference` for this instead of `Base.RefValue`? Someone's code could conceivably be using `Base.RefValue`s for other reasons, and then we'd need to add special handling in `object_and_preserve` to make sure it doesn't get stripped on the load.
Adding special handling wouldn't be that bad, so we could replace `Reference` with `Base.RefValue`.
```julia
julia> using ManualMemory

julia> x = zeros(UInt, 100);

julia> tref = ManualMemory.Reference((1.0,2.0,3.0))
ManualMemory.Reference{Tuple{Float64, Float64, Float64}}((1.0, 2.0, 3.0))

julia> ManualMemory.store!(Base.unsafe_convert(Ptr{typeof(tref)}, pointer(x)), tref)

julia> ManualMemory.load(Base.unsafe_convert(Ptr{typeof(tref)}, pointer(x)))
(1.0, 2.0, 3.0)

julia> typeof(ans) # not a Reference
Tuple{Float64, Float64, Float64}
```

It'd probably also make sense to move `object_and_preserver` here, I just wanted to turn `Array`s into `PtrArray`s, which are defined there (and depend on `ArrayInterface`, adding some more dependencies than we probably want to bring in here).

> Would it make sense to have `Reference{T}()` defined?

Sure, that'd probably be useful. `Ref{T}()` is.

**Comment by Tokazama** [comment]
> someone's code could conceivably be using Base.RefValues for other reasons

Got it. I can make a PR for for `Reference{T}()`.

**Comment by Tokazama** [comment]
Addressed with #7
